csr/csr.go

```diff
@@ -92,22 +96,34 @@ func VerifyCSR(ctx context.Context, csr *x509.CertificateRequest, maxNames int,
 	return nil
 }
 
-// normalizeCSR deduplicates and lowers the case of dNSNames and the subject CN.
-// It will also hoist a dNSName into the CN if it is empty.
-func normalizeCSR(csr *x509.CertificateRequest) {
-	if csr.Subject.CommonName == "" {
-		var forcedCN string
-		// Promote the first SAN that is less than maxCNLength (if any)
-		for _, name := range csr.DNSNames {
-			if len(name) <= maxCNLength {
-				forcedCN = name
-				break
-			}
+type names struct {
+	SANs []string
+	CN   string
+}
+
+// NamesFromCSR deduplicates and lower-cases the Subject Common Name and Subject
+// Alternative Names from the CSR. It guarantees that the SANs include the CN.
+// It also enforces various conditions on the CN, based on feature flags.
+func NamesFromCSR(csr *x509.CertificateRequest) names {
+	sans := csr.DNSNames
+	if csr.Subject.CommonName != "" {
+		sans = append(sans, csr.Subject.CommonName)
+	}
+	sans = core.UniqueLowerNames(sans)
+
+	if !features.Enabled(features.SetCommonName) {
+		return names{SANs: sans}
+	}
+
+	if csr.Subject.CommonName != "" {
+		return names{SANs: sans, CN: strings.ToLower(csr.Subject.CommonName)}
```

Comment:
```suggestion
		cn := strings.ToLower(csr.Subject.CommonName)
		return names{SANs: sans, CN: cn}
```

Reply:
Having small readable return lines is good, you're right. But in this case I'm prioritizing a different style guideline: don't create variables that you're only going to use once. Storing a value in a variable suggests to the reader that it _might_ be re-used far away from where it was created -- this increases mental load on the reader, who now has to keep an eye out for that variable showing up again elsewhere. Inlining this one function call into the return makes it clear that the value is never used elsewhere.